File: 02/02-2.py
from collections import Counter
from sys import exit

if __name__ == "__main__":
    valid = 0
    with open("input", "r") as o:
        for line in o.readlines():
            char_range, char, password = line.split(" ")

            char = char[:-1]  # remove :
            password = password.strip()  # remove newline

            first_position, second_position = [int(x) for x in char_range.split("-")]

            first_present = password[first_position - 1] == char
            second_present = password[second_position - 1] == char

            if not all([first_present, second_present]) and (
                first_present or second_present
            ):
                valid += 1

    print("Valid passwords: ", valid)

File: 2020/02/02-2.go
package main

import (
	"fmt"
	"strconv"
	"strings"

	. "../helpers"
)

func main() {
	var lines []string
	var err error
	if lines, err = ReadInput("./input"); err == nil {
		fmt.Printf("something bad", err)
	}

	validPasswords := 0

	for _, line := range lines {
		split_line := strings.Split(line, " ")

		charRange := split_line[0]
		keyChar := []rune(strings.Replace(split_line[1], ":", "", -1))
		password := strings.Trim(split_line[2], " \n")

		splitCharRange := strings.Split(charRange, "-")
		firstPosition, _ := strconv.Atoi(splitCharRange[0])
		secondPosition, _ := strconv.Atoi(splitCharRange[1])

		first := rune(password[firstPosition-1]) == keyChar[0]
		second := rune(password[secondPosition-1]) == keyChar[0]

		if !(first && second) && (first || second) {
			validPasswords++
		}

	}
	fmt.Printf("Valid Passwords: %d", validPasswords)
}
